File: packages/guardiandb-odm/src/query.ts
import { GuardianDBError } from "./errors.js";
import { deepEqual, getPath, setPath, unsetPath } from "./utils.js";
import type { Document, Query, UpdateOperations } from "./types.js";

export function matchesQuery<T extends Document>(document: T, query: Query<T>): boolean {
  for (const [field, condition] of Object.entries(query)) {
    if (field === "$and" || field === "$or" || field === "$nor") {
      if (!Array.isArray(condition)) {
        throw new GuardianDBError("INVALID_QUERY", `${field} expects an array`);
      }
      const clauses = condition as Query<T>[];
      if (field === "$and" && !clauses.every((clause) => matchesQuery(document, clause))) return false;
      if (field === "$or" && !clauses.some((clause) => matchesQuery(document, clause))) return false;
      if (field === "$nor" && clauses.some((clause) => matchesQuery(document, clause))) return false;
      continue;
    }
    if (field.startsWith("$")) {
      throw new GuardianDBError("INVALID_QUERY", `Unsupported logical operator '${field}'`);
    }
    if (!matchesCondition(getPath(document, field), condition)) {
      return false;
    }
  }
  return true;
}

export function applyUpdate<T extends Document>(
  document: T,
  operations: UpdateOperations<T>,
  immutableFields: ReadonlySet<string>,
): boolean {
  const keys = Object.keys(operations);
  if (keys.some((key) => !key.startsWith("$"))) {
    throw new GuardianDBError(
      "INVALID_UPDATE",
      "Replacement updates are not supported; use operators such as $set",
    );
  }

  let changed = false;
  if (operations.$set !== undefined) {
    for (const [path, value] of Object.entries(operations.$set)) {
      assertMutable(path, immutableFields);
      changed = setPath(document, path, value) || changed;
    }
  }
  if (operations.$unset !== undefined) {
    for (const path of Object.keys(operations.$unset)) {
      assertMutable(path, immutableFields);
      changed = unsetPath(document, path) || changed;
    }
  }
  if (operations.$inc !== undefined) {
    for (const [path, increment] of Object.entries(operations.$inc)) {
      assertMutable(path, immutableFields);
      if (typeof increment !== "number" || !Number.isFinite(increment)) {
        throw new GuardianDBError("INVALID_UPDATE", `$inc value for '${path}' must be numeric`);
      }
      const current = getPath(document, path);
      if (current !== undefined && typeof current !== "number") {
        throw new GuardianDBError("INVALID_UPDATE", `$inc target '${path}' must be numeric`);
      }
      changed = setPath(document, path, (current ?? 0) + increment) || changed;
    }
  }

  for (const key of keys) {
    if (key !== "$set" && key !== "$unset" && key !== "$inc") {
      throw new GuardianDBError("INVALID_UPDATE", `Unsupported update operator '${key}'`);
    }
  }
  return changed;
}

function matchesCondition(actual: unknown, condition: unknown): boolean {
  if (isOperatorObject(condition)) {
    for (const [operator, operand] of Object.entries(condition)) {
      switch (operator) {
        case "$eq":
          if (!equalWithArraySemantics(actual, operand)) return false;
          break;
        case "$ne":
          if (equalWithArraySemantics(actual, operand)) return false;
          break;
        case "$gt":
          if (!(compare(actual, operand) > 0)) return false;
          break;
        case "$gte":
          if (!(compare(actual, operand) >= 0)) return false;
          break;
        case "$lt":
          if (!(compare(actual, operand) < 0)) return false;
          break;
        case "$lte":
          if (!(compare(actual, operand) <= 0)) return false;
          break;
        case "$in":
          if (!Array.isArray(operand)) throw invalidOperand("$in", "an array");
          if (!operand.some((candidate) => equalWithArraySemantics(actual, candidate))) return false;
          break;
        case "$nin":
          if (!Array.isArray(operand)) throw invalidOperand("$nin", "an array");
          if (operand.some((candidate) => equalWithArraySemantics(actual, candidate))) return false;
          break;
        case "$exists":
          if (typeof operand !== "boolean") throw invalidOperand("$exists", "a boolean");
          if ((actual !== undefined) !== operand) return false;
          break;
        case "$size":
          if (typeof operand !== "number" || !Number.isInteger(operand) || operand < 0) {
            throw invalidOperand("$size", "a non-negative integer");
          }
          if (!Array.isArray(actual) || actual.length !== operand) return false;
          break;
        default:
          throw new GuardianDBError("INVALID_QUERY", `Unsupported field operator '${operator}'`);
      }
    }
    return true;
  }
  return equalWithArraySemantics(actual, condition);
}

function isOperatorObject(value: unknown): value is Record<string, unknown> {
  return (
    value !== null &&
    typeof value === "object" &&
    !Array.isArray(value) &&
    Object.keys(value).some((key) => key.startsWith("$"))
  );
}

function equalWithArraySemantics(actual: unknown, expected: unknown): boolean {
  return deepEqual(actual, expected) || (Array.isArray(actual) && actual.some((value) => deepEqual(value, expected)));
}

function compare(left: unknown, right: unknown): number {
  if (typeof left === "number" && typeof right === "number") return left - right;
  if (typeof left === "string" && typeof right === "string") return left.localeCompare(right);
  return Number.NaN;
}

function invalidOperand(operator: string, expected: string): GuardianDBError {
  return new GuardianDBError("INVALID_QUERY", `${operator} expects ${expected}`);
}

function assertMutable(path: string, immutableFields: ReadonlySet<string>): void {
  for (const field of immutableFields) {
    if (path === field || path.startsWith(`${field}.`)) {
      throw new GuardianDBError("IMMUTABLE_FIELD", `Immutable field '${field}' cannot be updated`, {
        field,
      });
    }
  }
}

File: src/odm/query.rs
use crate::odm::error::{OdmError, Result};
use serde_json::{Map, Value};
use std::cmp::Ordering;

/// Resolves a dotted field path (e.g. `"a.b.c"`) within a JSON document,
/// returning a reference to the nested value if every segment exists.
pub(crate) fn value_at_path<'a>(document: &'a Value, path: &str) -> Option<&'a Value> {
    let mut current = document;
    for segment in path.split('.') {
        current = current.as_object()?.get(segment)?;
    }
    Some(current)
}

/// Evaluates a MongoDB-style query object against a document, returning whether
/// the document matches. Supports logical operators (`$and`/`$or`/`$nor`) and
/// per-field operators (`$eq`, `$gt`, `$in`, `$exists`, ...).
pub(crate) fn matches_query(document: &Value, query: &Value) -> Result<bool> {
    let query = query
        .as_object()
        .ok_or_else(|| OdmError::InvalidQuery("query must be a JSON object".to_string()))?;
    matches_object(document, query)
}

fn matches_object(document: &Value, query: &Map<String, Value>) -> Result<bool> {
    for (field, condition) in query {
        match field.as_str() {
            "$and" => {
                let clauses = condition
                    .as_array()
                    .ok_or_else(|| OdmError::InvalidQuery("$and expects an array".to_string()))?;
                for clause in clauses {
                    if !matches_query(document, clause)? {
                        return Ok(false);
                    }
                }
            }
            "$or" => {
                let clauses = condition
                    .as_array()
                    .ok_or_else(|| OdmError::InvalidQuery("$or expects an array".to_string()))?;
                let mut matched = false;
                for clause in clauses {
                    if matches_query(document, clause)? {
                        matched = true;
                        break;
                    }
                }
                if !matched {
                    return Ok(false);
                }
            }
            "$nor" => {
                let clauses = condition
                    .as_array()
                    .ok_or_else(|| OdmError::InvalidQuery("$nor expects an array".to_string()))?;
                for clause in clauses {
                    if matches_query(document, clause)? {
                        return Ok(false);
                    }
                }
            }
            key if key.starts_with('$') => {
                return Err(OdmError::InvalidQuery(format!(
                    "unsupported logical operator `{key}`"
                )));
            }
            _ => {
                let actual = value_at_path(document, field);
                if !matches_condition(actual, condition)? {
                    return Ok(false);
                }
            }
        }
    }
    Ok(true)
}

fn matches_condition(actual: Option<&Value>, condition: &Value) -> Result<bool> {
    if let Some(operators) = condition.as_object()
        && operators.keys().any(|key| key.starts_with('$'))
    {
        for (operator, operand) in operators {
            let matched = match operator.as_str() {
                "$eq" => actual.is_some_and(|value| values_equal(value, operand)),
                "$ne" => actual.is_none_or(|value| !values_equal(value, operand)),
                "$gt" => compare(actual, operand, |ordering| ordering == Ordering::Greater),
                "$gte" => compare(actual, operand, |ordering| ordering != Ordering::Less),
                "$lt" => compare(actual, operand, |ordering| ordering == Ordering::Less),
                "$lte" => compare(actual, operand, |ordering| ordering != Ordering::Greater),
                "$in" => {
                    let values = operand.as_array().ok_or_else(|| {
                        OdmError::InvalidQuery("$in expects an array".to_string())
                    })?;
                    actual.is_some_and(|value| {
                        values
                            .iter()
                            .any(|candidate| values_equal(value, candidate))
                    })
                }
                "$nin" => {
                    let values = operand.as_array().ok_or_else(|| {
                        OdmError::InvalidQuery("$nin expects an array".to_string())
                    })?;
                    actual.is_none_or(|value| {
                        values
                            .iter()
                            .all(|candidate| !values_equal(value, candidate))
                    })
                }
                "$exists" => {
                    let expected = operand.as_bool().ok_or_else(|| {
                        OdmError::InvalidQuery("$exists expects a boolean".to_string())
                    })?;
                    actual.is_some() == expected
                }
                "$size" => {
                    let expected = operand.as_u64().ok_or_else(|| {
                        OdmError::InvalidQuery("$size expects a non-negative integer".to_string())
                    })? as usize;
                    actual
                        .and_then(Value::as_array)
                        .is_some_and(|items| items.len() == expected)
                }
                other => {
                    return Err(OdmError::InvalidQuery(format!(
                        "unsupported field operator `{other}`"
                    )));
                }
            };
            if !matched {
                return Ok(false);
            }
        }
        return Ok(true);
    }

    Ok(actual.is_some_and(|value| values_equal(value, condition)))
}

/// Equality used by query operators: a direct match, or — when the actual value
/// is an array — a match against any of its elements (array containment).
fn values_equal(actual: &Value, expected: &Value) -> bool {
    if actual == expected {
        return true;
    }
    actual
        .as_array()
        .is_some_and(|items| items.iter().any(|item| item == expected))
}

/// Compares the actual and expected values (numbers or strings) and applies
/// `predicate` to the resulting ordering; returns `false` when uncomparable.
fn compare(
    actual: Option<&Value>,
    expected: &Value,
    predicate: impl FnOnce(Ordering) -> bool,
) -> bool {
    let Some(actual) = actual else {
        return false;
    };

    let ordering = match (actual, expected) {
        (Value::Number(left), Value::Number(right)) => left
            .as_f64()
            .zip(right.as_f64())
            .and_then(|(left, right)| left.partial_cmp(&right)),
        (Value::String(left), Value::String(right)) => Some(left.cmp(right)),
        _ => None,
    };
    ordering.is_some_and(predicate)
}
